lib/taiko.js

```diff
@@ -582,15 +632,18 @@ module.exports.closeTab = async targetUrl => {
       id: target.id,
     });
   }
-  if (!targetHandler.isMatchingUrl(others[0], currentUrl)) {
+  if (
+    !targetHandler.isMatchingUrl(others[0], currentUrl) &&
+    !targetHandler.isMatchingRegex(others[0], currentUrl)
+  ) {
     _client.removeAllListeners();
     _client = null;
     await connect_to_cri(others[0]);
     await dom.getDocument();
   }
   let message = targetUrl
-    ? `Closed all tabs with URL ${targetUrl}`
-    : `Closed current tab with URL ${closedTabUrl}`;
+    ? `Closed all tabs with${isRegexText} URL ${targetUrl}`
+    : `Closed current tab with${isRegexText} URL ${closedTabUrl}`;
   descEvent.emit('success', message);
 };
 
```

Comment:
* The argument check for `targetUrl` can be removed as it happens in `getCriTargets`
* Lot of if conditions to handle the display message. It's better to make it generic 

Here are some modifications I did to the whole call chain the changes will be similar for `switchTo`

```
module.exports.closeTab = async targetUrl => {
  validate();

  const { matching, others } = await targetHandler.getCriTargets(
    targetUrl,
    currentHost,
    currentPort,
  );
  if (!others.length) {
    await _closeBrowser();
    descEvent.emit('success', 'Closing last target and browser.');
    return;
  }
  if (!matching.length) {
    throw new Error('No target with given URL/Title/Regex found.');
  }
  let currentUrl = await currentURL();
  let closedTabUrl;
  for (let target of matching) {
    closedTabUrl = target.url;
    await cri.Close({
      host: currentHost,
      port: currentPort,
      id: target.id,
    });
  }
  if (
    !targetHandler.isMatchingUrl(others[0], currentUrl) &&
    !targetHandler.isMatchingRegex(others[0], currentUrl)
  ) {
    _client.removeAllListeners();
    _client = null;
    await connect_to_cri(others[0]);
    await dom.getDocument();
  }
  let message = `Closed tab(s) matching ${targetUrl || closedTabUrl }`
  descEvent.emit('success', message);
};
```

```
const getCriTargets = async function(targetObject, host, port) {
  let targets = await cri.List({ host: host, port: port });
  let pages = targets.filter(target => target.type === 'page');
  let response = {
    matching: pages.slice(0, 1),
    others: pages.slice(1),
  };
  if (targetObject) {
    response = { matching: [], others: [] };
    for (const target of targets) {
      if (
        (targetObject instanceof RegExp && isMatchingRegex(target, targetObject)) ||
        isMatchingUrl(target, targetObject)
      ) {
        response.matching.push(target);
      } else {
        response.others.push(target);
      }
    }
  }
  return response;
};
```

```
const isMatchingRegex = function(target, targetRegex) {
  const parsedUrl = url.parse(target.url, true);
  const host = parsedUrl.host ? parsedUrl.host : '';
  const urlPath = path.join(
    host,
    trimCharLeft(parsedUrl.pathname, '/'),
  );
  const urlHrefPath = parsedUrl.protocol
    .concat('//')
    .concat(trimCharLeft(host, '/'));

  return (
    target.title.match(targetRegex) ||
    urlPath.match(targetRegex) ||
    parsedUrl.href.match(targetRegex) ||
    urlHrefPath.match(targetRegex)
  );
};

const isMatchingUrl = function(target, targetUrl) {
  const parsedUrl = url.parse(target.url, true);
  const parsedTargetUrl = url.parse(targetUrl, true);

  const targetHost = parsedTargetUrl.host ? parsedTargetUrl.host : '';
  const host = parsedUrl.host ? parsedUrl.host : '';

  const targetUrlPath = path.join(
    targetHost,
    trimCharLeft(parsedTargetUrl.pathname, '/'),
  );
  const urlPath = path.join(
    host,
    trimCharLeft(parsedUrl.pathname, '/'),
  );

  target.url = handleUrlRedirection(target.url);
  return (
    target.title === escapeHtml(targetUrl) ||
    urlPath === (targetUrlPath || targetUrl) ||
    parsedUrl.href === targetUrl
  );
};
```

Notice that there are no parameter condition checks in `isMatchingUrl` and `isMatchingRegex` The responsibility can be at `getCRITargets` i.e. these lines

```
     if (
        (targetObject instanceof RegExp && isMatchingRegex(target, targetObject)) ||
        isMatchingUrl(target, targetObject)
      )
```

One of the modified test will be as follows. 

```
it('should close the current tab and switch to last active target if no url given', async () => {
    _targets.matching.push({
      id: '1',
      type: 'page',
      url: 'https://flipkart.com',
    });
    _targets.others.push({
      id: '2',
      type: 'page',
      url: 'https://flipkart.com',
    });
    _targets.others.push({
      id: '3',
      type: 'page',
      url: 'https://amazon.com',
    });
    currentURL = 'https://amazon.com';
    _isMatchUrl = false;
    _isMatchRegex = false;
    let validatePromise = validateEmitterEvent(
      'success',
      'Closed tab(s) matching https://flipkart.com',
    );
    await taiko.closeTab();
    await validatePromise;
    expect(currentTarget.url).to.be.eql('https://flipkart.com');
  });
```

Reply:
I see you have added this condition 
`if (
        (targetObject instanceof RegExp) && isMatchingRegex(target, targetObject)) ||
        isMatchingUrl(target, targetObject)
      )`

But this doesn't work always.
Suppose that I pass something like,
`
openBrowser()
goto('google.com')
openTab('github.com')
switchTo(/Go*gle/)
`
The stacktrace would be something like.
1. Go to `isMatchingRegex("github.com",/Go*gle/)`.
  The above statement will return false. and hence it will go and try executing isMatchingUrl().
2. Go to `isMatchingUrl('github.com',/Go*gle/)`
  For this method, the below line will cause exception.
`const parsedTargetUrl = url.parse(targetUrl, true);`

We are also looping through all the tabs open in the code. Hence, the check in both isMatchingUrl() & isMatchingRegex() is required. Please let me know what you think.